### DISCUSSION TITLE
i18n: not able to generate pages for other language

### DISCUSSION THREAD

**Discussion by Kvaibhav01** [vercel/next.js#43078] [Help] [answered]
Hello!

I want to generate dynamic pages for my i18n routes. For this here's what I have done till now:

1. In my _index.js page_ I have given link to the dynamic `href`:

```jsx
<Link href={{ pathname: '/news/[slug]', query: { slug: news.slug } }} locale = {locale}>
   <h1>{news.title}</h1>
</Link>
```

2. In the _/news/[slug].js_ file:

```jsx
export const getStaticPaths = async () => {
  const res = await fetch('http://localhost:3001/');
  const data = await res.json();

  const paths = data.map((news) => {
    return {
      params: {
        slug: news.slug,
      },
    };
  });

  return {
    paths,
    fallback: false,
  };
};


export const getStaticProps = async (context) => {
  const newsSlug = context.params.slug;
  const currentLocale = context.locale; 

  const res = await fetch(`http://localhost:3001/${currentLocale}/${newsSlug}`);
  const data = await res.json();

  return {
    props: {
      data,
    },
  };
};
```

3. Then I have a `LanguageSwitcher` component which is a `<select>` dropdown with the `onChange` I am using `router.push()` as:

```jsx
      <select
        onChange={(e) =>
          router.push(router.pathname, router.pathname, {
            locale: e.target.value,
          })
        }
      >
        <option value='en'>English</option>
        <option value='es'>Español</option>
      </select>
```

Now, when I go to the index route or to the dynamic route, the default `en` language content is rendered well. But the issue is when I change the language to `es` using the `LanguageSwitcher` in the dynamic route it throws an error:

> Error: The provided `href` (/news/[slug]) value is missing query values (slug) to be interpolated properly. Read more: https://nextjs.org/docs/messages/href-interpolation-failed

I read the interpolation failed link, it shows an example where they are adding the right `query` values in the `Link` but I have already done that above in _index.js_.

Also, when I first change the language to `es` on the index route and then visit the dynamic page like: `http://localhost:3000/es/news/my-news-slug`, then it shows the 404: This page couldn't be found page.

I debugged to know that the `currentLocale` isn't changing to `es` or some other locale via the dropdown. Due to this, the dynamic pages for the `es` locale are not built.

How can I solve this issue? Thanks!

Note that the backend API with `http://localhost:3001/es/my-news-slug` works correctly and gets back the data required.

**Comment by icyJoseph** [comment, marked as answer]
I think in the `router.push` done in your switch you need to include the query as well:

```js
  router.push({
          pathname: router.pathname,
          query: router.query,
        }, null, {locale: e.target.value})
```

Pass null as second argument so you can skip trying reconstructing the url all the time, the first argument is enough, and well, you need the third to tell Next.js to swap the locale.

That should take care of changing your locale inside the `news/[slug]` page. 

Do remember that when you push, you override not only the path but also any queries present, and that to keep the `[slug]` you gotta forward the query object contained in `next/router`.

For the second problem, in your `getStaticPaths` you also gotta tell `Next.js` what locales to build, https://nextjs.org/docs/api-reference/data-fetching/get-static-paths#paths:

> Separate of the params object a locale field can be returned when [i18n is configured](https://nextjs.org/docs/advanced-features/i18n-routing), which configures the locale for the path being generated.


```ts
export const getStaticPaths: GetStaticPaths<{ page: string }> = ({
  locales,
}) => {
  // locales are not set, maybe you can throw or something
  if (!locales) return paths;

  const pathsWithLocale = [{ slug: "1" }, { slug: "2" }].flatMap((news) => {
    return locales?.map((locale) => ({
      params: {
        slug: news.slug,
      },
      locale,
    }));
  });

  return {
    paths: pathsWithLocale,
    fallback: false,
  };
};
```

**Reply by Kvaibhav01** [reply]
Hi @icyJoseph thanks for your reply. I can't understand this statement:

> Pass null as second argument so you can skip trying reconstructing the url all the time

The second argument for `router.push` is `as` right? That is what is shown in the URL?

**Reply by icyJoseph** [reply]
Yeah, but if you don't pass anything the router should be able to resolve it, you almost never have to pass it right? 

Normally you call `router.push("/something")` but since we need to reach to the 3rd position argument, and in JS you can't just say `router.push("/something", , {locale: 'en'})`, then you gotta drop something in place. `undefined` also kinda works.

Take a look here, https://nextjs.org/docs/routing/shallow-routing:

```js
  useEffect(() => {
    // Always do navigations after the first render
    router.push('/?counter=10', undefined, { shallow: true })
  }, [])
```

**Reply by Kvaibhav01** [reply]
Okay got it. It's more clear now about shallow routing.

But on the `getStaticPaths` code you shared above, it will generate only the paths of `http://localhost:3000/news/1` and `http://localhost:3000/news/2` as we provided `{ slug: '1' }, { slug: '2' }` in the array. How can I add the dynamic slug value here so that all the matching slugs from the JSON data are used to generate corresponding pages?

**Reply by icyJoseph** [reply]
I mean, you have to adapt the code to your situation, I think it'd be like this though:

```js
export const getStaticPaths = async ({ locales }) => {
  const res = await fetch("http://localhost:3001/");
  const data = await res.json();

  const paths = data.flatMap((news) => {
    return locales.map((locale) => ({
      params: {
        slug: news.slug,
      },
      locale,
    }));
  });

  return {
    paths,
    fallback: false,
  };
};
```

And regarding shallow routing that was just an example to show that you can "skip" the second parameter to router push, so that you can pass the locale. You don't need shallow routing in this case.

**Reply by icyJoseph** [reply]
You'll know you have it when the build result prints something like:

```
Route (pages)                              Size     First Load JS
┌ ○ /                                      6.54 kB        86.4 kB
├   /_app                                  0 B            79.9 kB
├ ○ /404                                   212 B          80.1 kB
├ λ /api/hello                             0 B            79.9 kB
├ ● /posts/[page] (ISR: 10 Seconds)        509 B          80.4 kB
├   ├ /en/posts/0
├   ├ /es/posts/0
├   ├ /en/posts/1
├   └ [+5 more paths]
# cut for readability
```

Notice how my posts are prefixed with locales.

**Reply by Kvaibhav01** [reply]
Thanks a lot! It works now! I learned about that we need to pass `locale` when building pages.
